src/primitive.ts

```diff
@@ -36,141 +36,167 @@ let LEADING_ZERO = /^[+-]?0[0-9_]/
+      }
+    } else if ((c < '\x20' && c !== '\t') || c === '\x7f') {
+      throw new TomlError('control characters are not allowed in strings', {
+        toml: str,
+        ptr: ptr - 1,
+      })
+    }
+
+    if (isEscape) {
+      isEscape = false
+      if (c === 'u' || c === 'U') {
+        // Unicode escape
+        let code = str.slice(ptr, (ptr += c === 'u' ? 4 : 8))
+        if (!ESCAPE_REGEX.test(code)) {
+          throw new TomlError('invalid unicode escape', {
+            toml: str,
+            ptr: tmp,
+          })
+        }
+
+        try {
+          parsed += String.fromCodePoint(parseInt(code, 16))
+        } catch {
+          throw new TomlError('invalid unicode escape', {
+            toml: str,
+            ptr: tmp,
+          })
+        }
+      } else if (isMultiline && (c === '\n' || c === ' ' || c === '\t' || c === '\r')) {
+        // Multiline escape
+        ptr = skipVoid(str, ptr - 1, true)
+        if (str[ptr] !== '\n' && str[ptr] !== '\r') {
+          throw new TomlError('invalid escape: only line-ending whitespace may be escaped', {
+            toml: str,
+            ptr: tmp,
+          })
+        }
+        ptr = skipVoid(str, ptr)
+      } else if (c in ESC_MAP) {
+        // Classic escape
+        parsed += ESC_MAP[c as keyof typeof ESC_MAP]
+      } else {
+        throw new TomlError('unrecognized escape sequence', {
+          toml: str,
+          ptr: tmp,
+        })
+      }
+
+      sliceStart = ptr
+    } else if (!isLiteral && c === '\\') {
+      tmp = ptr - 1
+      isEscape = true
+      parsed += str.slice(sliceStart, tmp)
+    }
+  }
+
+  return parsed + str.slice(sliceStart, endPtr - 1)
 }
 
-export function parseValue (value: string, toml: string, ptr: number): boolean | number | TomlDate {
-	// Constant values
-	if (value === 'true') return true
-	if (value === 'false') return false
-	if (value === '-inf') return -Infinity
-	if (value === 'inf' || value === '+inf') return Infinity
-	if (value === 'nan' || value === '+nan' || value === '-nan') return NaN
-
-	if (value === '-0') return 0 // Avoid FP representation of -0
-
-	// Numbers
-	let isInt
-	if ((isInt = INT_REGEX.test(value)) || FLOAT_REGEX.test(value)) {
-		if (LEADING_ZERO.test(value)) {
-			throw new TomlError('leading zeroes are not allowed', {
-				toml: toml,
-				ptr: ptr
-			})
-		}
-
-		let numeric = +(value.replace(/_/g, ''))
-		if (isNaN(numeric)) {
-			throw new TomlError('invalid number', {
-				toml: toml,
-				ptr: ptr
-			})
-		}
-
-		if (isInt && !Number.isSafeInteger(numeric)) {
-			throw new TomlError('integer value cannot be represented losslessly', {
-				toml: toml,
-				ptr: ptr
-			})
-		}
-
-		return numeric
-	}
-
-	let date = new TomlDate(value)
-	if (!date.isValid()) {
-		throw new TomlError('invalid value', {
-			toml: toml,
-			ptr: ptr
-		})
-	}
-
-	return date
+export function parseValue(
+  value: string,
+  toml: string,
+  ptr: number,
+): boolean | number | TomlDate | BigInt {
```

Comment:
[Don't use boxed types](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#number-string-boolean-symbol-and-object). The doc doesn't mention BigInt but it also applies to it. Use lowercase `bigint`.